src/rt-app_types.h

```diff
@@ -150,6 +150,12 @@ typedef struct _rtapp_resource_t {
 	char *name;
 } rtapp_resource_t;
 
+typedef struct _taskgroup_data_t {
+	char name[FILENAME_MAX];
+	/* Start of taskgroup part which is under control of rt-app run. */
+	int offset;
```

Comment:
If we really wanted to avoid allocating lots of FILENAME_MAX length structs, we could instead put offset (and length) first and allocate right-sized objects since we only use pointers to these in the thread/phase objects.

It might make using a single array of them more difficult, so we might also want to think about making them into a linked list. Something like?

```
typedef struct _taskgroup_data_t {
		/* housekeeping */
		struct _taskgroup_data_t *next; 
￼		/* Start of taskgroup part which is under control of rt-app run. */
￼		int offset;
		int name_len;
￼		char *name;
} taskgroup_data_t;
```

Reply:
This can be implemented as well. I will dynamically only allocate the memory necessary for each taskgroup name.

Reply:
I played with different possibilities to allocate the taskgroups and the most suitable is a malloced/realloced array of taskgroup_data_t for which the name member is allocated dynamically as well.

```
typedef struct _taskgroup_data_t {
        char *name;
        int offset;
} taskgroup_data_t;
```

This would allow that threads and phases use the same taskgroup_data_t in case they are configured with the same 'taskgroup' json strings.

And we would only allocate the number bytes necessary for the taskgroup string and not FILENAME_MAX.

The offset mechanism (taskgroup_data_t member offset) is useful since it allows to keep cgroups which have been set up outside an rt-app run and are part of a 'taskgroup' json string.

E.g. a system is configured like this before an rt-app run:

```
/sys/fs/cgroup/cpu,cpuacct$ tree -d
.
├── tg1
└── tg2
    └── tg21
```

and the rt-app sets up '/tg1/tg11' and '/tg2/tg21/tg211'

```
.
├── tg1
│   └── tg11
└── tg2
    └── tg21
        └── tg211

```
The offset mechanism guarantees that after the rt-app run, the system will still look like:

```
.
├── tg1
└── tg2
    └── tg21
```

Moreover, with the offset mechanism, the taskgroup array does not have to be sorted (alphabetical sort of the names) because the offset mechanism includes cgroup name part which have been set up by taskgroups of other threads or phases of the rt-app run.

Reply:
I abandoned the realloc idea and went for a array with 32 entries instead. With realloc, the pointer to the array changes so that I would need to iterate twice to the config data (1. to allocate/reallocate the needed array entries and 2. to assign the pointers to it). In the unlikely case that an rt-app configuration would need more than 32 different taskgroup entries, the tool could be easily rebuild with a higher value.